xls/modules/fp_trig_reduce_64_test.cc

```diff
@@ -0,0 +1,261 @@
+typedef struct {
+  // Fields are reversed relative to xls
+  // because earlier fields in an xls tuple
+  // occupy higher bits.
+  double fraction;
+  uint8_t j;
+} ResultT;
+
+using PackedFloat64 = PackedTupleView<PackedBitsView<1>, PackedBitsView<11>, PackedBitsView<52>>;
+using PackedU3Float64 = PackedTupleView<PackedBitsView<3>, PackedFloat64>;
+
+double FlushSubnormals(double value) {
+  if (std::fpclassify(value) == FP_SUBNORMAL) {
+    return 0;
+  }
+  return value;
+}
+
+bool ZeroOrSubnormal(double value) {
+  return value == 0 || std::fpclassify(value) == FP_SUBNORMAL;
+}
+
+// Cast from uint64_t to double.
+double unflatten(uint64_t in) {
+  return absl::bit_cast<double>(in);
+}
+
+// Cast from double to uint64_t
+uint64_t flatten(double in) {
+  return absl::bit_cast<uint64_t>(in);
+}
+
+// Return the number of leading zeros in num;
+uint64_t clz(uint64_t num) {
+  uint64_t leading_zeros = 0;
+  for(uint64_t mask = static_cast<uint64_t>(1) << 63;
+      !(mask & num);
+      mask = mask >> 1) {
+    leading_zeros++;
+  }
+  return leading_zeros;
+}
+
+// Generates double with reasonably unformly random bit patterns.
+double IndexToInput(uint64_t index) {
+  thread_local absl::BitGen bitgen;
+  double a = absl::Uniform(bitgen, 0u, std::numeric_limits<uint64_t>::max());
+  return a;
+}
+
+ResultT ComputeExpected(double x) {
+  // Bits of 4/Pi.
+  constexpr uint64_t MPI4[] = {
+    0x0000000000000001,
+    0x45f306dc9c882a53,
+    0xf84eafa3ea69bb81,
+    0xb6c52b3278872083,
+    0xfca2c757bd778ac3,
+    0x6e48dc74849ba5c0,
+    0x0c925dd413a32439,
+    0xfc3bd63962534e7d,
+    0xd1046bea5d768909,
+    0xd338e04d68befc82,
+    0x7323ac7306a673e9,
+    0x3908bf177bf25076,
+    0x3ff12fffbc0b301f,
+    0xde5e2316b414da3e,
+    0xda6cfd9e4f96136e,
+    0x9e8c7ecd3cbfd45a,
+    0xea4f758fd7cbe2f6,
+    0x7a0e73ef14a525d4,
+    0xd7f6bf623f1aba10,
+    0xac06608df8f6d757,
+  };
+
+  constexpr uint64_t FP64_EXP_SHIFT = 52;
+  constexpr uint64_t FP64_EXP_BIAS = 1023;
+  constexpr uint64_t FP64_EXP_MASK = 0x7ff;
+
+
+  if (x < M_PI_4) {
+    return ResultT{.fraction = x, .j = static_cast<char>(0)};
+  }
+
+  // Extract out the integer and exponent such that,
+	// x = ix * 2 ** exp.
+  uint64_t ix = flatten(x);
+  int exp = int((ix>>FP64_EXP_SHIFT)&FP64_EXP_MASK) - FP64_EXP_BIAS - FP64_EXP_SHIFT;
+  ix  = ~(FP64_EXP_MASK << FP64_EXP_SHIFT) & ix;
+  ix |= uint64_t(1) << FP64_EXP_SHIFT;
+
+  // Use the exponent to extract the 3 appropriate uint64 digits from MPI4,
+  uint64_t digit = uint64_t(exp+61)/64;
+  uint64_t bitshift = uint64_t(exp+61)%64;
+
+  uint64_t z0 = (MPI4[digit] << bitshift) | 
+                  // It turns out that a shift of size >= type size (e.g. 64)
+                  // is undefined behavior in C. Bafflingly, uint64_t << 64
+                  // result in 0 if the shift amount is statically determined
+                  // but results in no shift if the shift ammount is dynamically
+                  // determined...
```

Comment:
I suspect this behavior is because in the case where the shift amount is statically known the shift is done by the compiler, in the case where it is not the processor does it, and evidently the processor masks out the high bits of the shift amount.

Reply:
Yes, that is my assumption.  Removed this part of the comment to avoid unnecessary confusion.